Avoid generating as_strided for alaising views in auto_functionalize_v2

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137149

during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

laithsakka committed:
avoid generating as_strided for alaising views in auto_functionalize_v2

[ghstack-poisoned]

laithsakka committed:
Update on "avoid generating as_strided for alaising views in auto_functionalize_v2"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


title, see unit tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we just pass the base.  (we can potentially call alias instead to be persisitance with weried case that check id inside custom op) not sure if we shall do that.

Those checks should not be inside auto_functionalized_dense to avoid generating guards in any case. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
> during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we just pass the base. (we can potentially call alias instead to be persisitance with weried case that check id inside custom op) not sure if we shall do that.

Can the Tensor ever be used later on? If so then we need to call alias to preserve the semantics (that the tensor is a view instead of the original tensor)

laithsakka commented:
> > during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
> > we just pass the base. (we can potentially call alias instead to be persisitance with weried case that check id inside custom op) not sure if we shall do that.
> 
> Can the Tensor ever be used later on? If so then we need to call alias to preserve the semantics (that the tensor is a view instead of the original tensor)

inside the custom op?

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +964,106 @@ def f(x):
+                        graph_inductor,
+                        """\
+def forward(self, arg0_1: "Sym(s0)", arg1_1: "f32[s0][1]cpu"):
+        foo_default = torch.ops.mylib.foo.default(arg1_1, arg1_1);  foo_default = None
+        copy_: "f32[s0][1]cpu" = torch.ops.aten.copy_.default(arg1_1, arg1_1);  copy_ = None
+        return (arg1_1, arg1_1)""",  # noqa: B950
```
There are two problems here:
1) the alias call should remain. Otherwise, we have object identity issues (notice how in the AOT graph, the returns do not have the same object identity. In the Inductor output, the returns have the same identity
2) (maybe this will go away when you fix (1)) the `torch.ops.aten.copy_.default(arg1_1, arg1_1)` seems like a no-op

laithsakka replied:
for (2) , I had a remove_copy function and was told to remove it since things will be optimized later on.

laithsakka replied:
(1) added a test and fixed the issue another bug down the road seems to fail the fail the test, skipping it for now
https://github.com/pytorch/pytorch/issues/137434.

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +964,106 @@ def f(x):
+        return (arg0_1, arg0_1)""",  # noqa: B950
+                        ignore_comments=True,
+                        ignore_empty_lines=True,
+                    )
```
can we do some sort of test with Dynamic shapes?
1) I expect arg0_1 to have symints as inputs. It doesn't look like it has symints as inputs. I'm not sure if that's because we're setting up the test wrong, or we are accidentally specializing
2) we should have a test that checks that we're not accidentally specializing the symints

zou3519 reviewed (changes requested):
Let's beef up the testing with dynamic shapes. Also, we shouldn't delete the as_strided calls, they should be turned into alias calls.

laithsakka commented:
so for an example like this one:
```
@torch.compile(fullgraph=True, dynamic=True, backend="inductor")
def f(a):
   b = torch.ops.aten.alias (a)
   foo(b)
```
I printed guards with and with out this change, 
We do not add new guards
```
--- out1.txt	2024-10-03 17:50:38.733158923 -0700
+++ out2.txt	2024-10-03 17:51:30.878880262 -0700
@@ -10,18 +10,18 @@
 | | +- NO_HASATTR: hasattr(L['a'], '_dynamo_dynamic_indices') == False           # b = torch.ops.aten.alias (a)  # example.py:15 in f
 | +- GuardManager: source=G, accessed_by=GlobalsGuardAccessor
 | | +- GuardManager: source=G['foo'], accessed_by=DictGetItemGuardAccessor(foo)
-| | | +- TYPE_MATCH: ___check_type_id(G['foo'], 82817104)                          # foo(b)  # example.py:16 in f
-| | | +- ID_MATCH: ___check_obj_id(G['foo'], 139978915343296)                    # foo(b)  # example.py:16 in f
+| | | +- TYPE_MATCH: ___check_type_id(G['foo'], 82816832)                          # foo(b)  # example.py:16 in f
+| | | +- ID_MATCH: ___check_obj_id(G['foo'], 140004880198592)                    # foo(b)  # example.py:16 in f
 | | | +- GuardManager: source=G['foo']._opoverload, accessed_by=GetAttrGuardAccessor(_opoverload)
-| | | | +- ID_MATCH: ___check_obj_id(G['foo']._opoverload, 139978915570320)        # return self._opoverload(*args, **kwargs)  # _library/custom_ops.py:669 in __call__
+| | | | +- ID_MATCH: ___check_obj_id(G['foo']._opoverload, 140004880425616)        # return self._opoverload(*args, **kwargs)  # _library/custom_ops.py:669 in __call__
 | | +- GuardManager: source=G['torch'], accessed_by=DictGetItemGuardAccessor(torch)
-| | | +- ID_MATCH: ___check_obj_id(G['torch'], 139978915439664)                  # b = torch.ops.aten.alias (a)  # example.py:15 in f
+| | | +- ID_MATCH: ___check_obj_id(G['torch'], 140004880294960)                  # b = torch.ops.aten.alias (a)  # example.py:15 in f
 | | | +- GuardManager: source=G['torch'].ops, accessed_by=GetAttrGuardAccessor(ops)
-| | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops, 139977039872464)              # b = torch.ops.aten.alias (a)  # example.py:15 in f
+| | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops, 140003014130592)              # b = torch.ops.aten.alias (a)  # example.py:15 in f
 | | | | +- GuardManager: source=G['torch'].ops.aten, accessed_by=GetAttrGuardAccessor(aten)
-| | | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops.aten, 139976863948912)         # b = torch.ops.aten.alias (a)  # example.py:15 in f
+| | | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops.aten, 140002829917840)         # b = torch.ops.aten.alias (a)  # example.py:15 in f
 | | | | | +- GuardManager: source=G['torch'].ops.aten.alias, accessed_by=GetAttrGuardAccessor(alias)
-| | | | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops.aten.alias, 139976680542928)   # b = torch.ops.aten.alias (a)  # example.py:15 in f
+| | | | | | +- ID_MATCH: ___check_obj_id(G['torch'].ops.aten.alias, 140002655949520)   # b = torch.ops.aten.alias (a)  # example.py:15 in f
 +- LAMBDA_GUARD: L['a'].size()[1] == L['a'].size()[0]  # duck sizing added this equality because these variables had the same size 2 (to avoid this specialization, set torch.fx.experimental._config.use_duck_shape = False)
 +- LAMBDA_GUARD: L['a'].stride()[0] == L['a'].size()[0]*L['a'].size()[2]  # (unknown source L['a'].stride()[0], please file a bug)
 +- LAMBDA_GUARD: L['a'].stride()[1] == L['a'].size()[2]  # (unknown source L['a'].stride()[1], please file a bug)
```
However I am unsure if there is other cases to worry about/test
also I am unsure what is the best way to test that specific peice of compiler code did not add guard.

laithsakka commented:
so we do add guards when we fail the check 
for example: 
```
@torch.compile(fullgraph=True, dynamic=True, backend="inductor")
def f(a):
   b = a[0]
   foo(b)
```
```
+- LAMBDA_GUARD: Ne(L['a'].size()[0], L['a'].size()[2])  # (_higher_order_ops/auto_functionalize.py:81 in write_single_view)
```
we do not add new guards for this
```
@torch.compile(fullgraph=True, dynamic=True, backend="inductor")
def f(a):
   b = torch.ops.aten.alias (a)
   foo(b)
```
because the two side of the symbol we are comparing are the same symbols. 
 
however what is the likely hood of actually failing the guards in next iterations, is your concern about the guard checking 
over head that we can avoid if we do this statically (capturing the alias call). or is your concern with potential recompilation.

laithsakka commented:
I see the point that "this guards is not needed in the case above, because we always know that it will is true" and we cant determine that statically,

but someone can counter that with " well if its always the same then we dont have to worry about recompilation"

I do not have strong opinion about landing this just sharing my thoughts, lmk what do you think

zou3519 commented:
The as_strided -> alias() change sounds good then, because it does not add new guards. Can you add additional tests to show that it doesn't invoke a recompile when the shape changes?

I'm just concerned about unnecessary recompiles.

For the slice change: I'm confused, why is it generating a guard that looks like `Ne(L['a'].size()[0], L['a'].size()[2]) `?

laithsakka commented:
> as_strided -> alias()

oh Ne(L['a'].size()[0], L['a'].size()[2])  is added by the alias change not by the slice change. I did not check anything on the other slice change yet. 
basically we fail the check 
``
     base.size() == tensor.size() 
```
and this cause the above failure.

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we just pass the base.  (we can potentially call alias instead to be persisitance with weried case that check id inside custom op) not sure if we shall do that.

Those checks should not be inside auto_functionalized_dense to avoid generating guards in any case. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka commented:
added recompilation tests, not sure if that covers your intention of :
```
we should also have a test where we create a tensor with unbacked symints, and then we take an alias of that, and then we mutate it.
```

also update the diff to generate alias, (added a test but it fail seems that soemwhere down the road inductor remove the alias ) see issue https://github.com/pytorch/pytorch/issues/137434

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we just pass the base.  (we can potentially call alias instead to be persisitance with weried case that check id inside custom op) not sure if we shall do that.

Those checks should not be inside auto_functionalized_dense to avoid generating guards in any case. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +966,219 @@ def f(x):
+        alias_default_1: "f32[s0][1]cpu" = torch.ops.aten.alias.default(arg1_1)
+        foo_default = torch.ops.mylib.foo.default(alias_default, alias_default_1);  \
+alias_default = alias_default_1 = foo_default = None
+        copy_: "f32[s0][1]cpu" = torch.ops.aten.copy_.default(arg1_1, arg1_1);  copy_ = None
```
To check... does inductor remove the copy_ during lowering?

laithsakka replied:
output code
```
def call(args):
    arg0_1, arg1_1, arg2_1 = args
    args.clear()
    s0 = arg0_1
    s1 = arg1_1
    assert_size_stride(arg2_1, (s0, s1), (s1, 1))
    # Topologically Sorted Source Nodes: [], Original ATen: []
    torch.ops.mylib.foo.default(arg2_1, arg2_1)
    return (arg2_1, arg2_1, )
```

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +966,219 @@ def f(x):
+                x.sin_()
+                y.sin_()
+
+            def f(x):
+                a = torch.ops.aten.alias.default(x)
```
can you make this something like...
```py
def f(x):
    a = torch.ops.aten.alias.default(x)
    b = x.clone()
    c = b.nonzero().float()
    d = c.alias()
    torch.ops.mylib.foo(a, d)
    return a, d
```
`d` is a Tensor with unbacked symints in the shape

laithsakka replied:
i will add a test for that i here and in the recompile test

laithsakka replied:
added they pass

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +966,219 @@ def f(x):
+    # that id(x) != id(base)
+    @torch._inductor.config.patch(enable_auto_functionalized_v2=True)
+    @unittest.skip(
+        reason="This test fails because something else in inductor optimize out the alias. issue #137434"
```
hmmm

zou3519 commented on test/inductor/test_auto_functionalize.py:
```diff
@@ -964,6 +966,219 @@ def f(x):
+            def not_eq_impl(x, y):
+                self.assertNotEqual(id(x), id(y))
+
+            def func(x):
+                a = torch.ops.aten.alias.default(x)
+                torch.ops.mylib.not_eq(a, x)
+
```
can you return `a`? and then assert that the input and output of `func` have different id identities.

Inductor doesn't match the tensor identity of intermediates of the function, but it should better match the identity of inputs/outputs of the compiled function.

laithsakka replied:
i will add another test that does that

laithsakka replied:
added that , that one pass.

zou3519 reviewed (approved):
code LGTM, but please see my suggestions for the test cases

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka committed:
Update on "Avoid generating as_strided for alaising views in auto_functionalize_v2"


during auto_functionalize_v2 if we encounter a view such that size() stride() and storage_offset() matches the base
we create a view that is regenerated by calling aten.alias instead of as_strided for better performance. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
